Addon API: Add experimental `page` addon type

ndelangen opened the pull request:
## What I did

- Introduce a new addon type: `page` (prefix with experimental for now)
- Simplify the API for this new addon type
- Make API for addon injection stricter
- Fix a bug where selection during preview initialization was ignored
- Make the default route `/` even if there's no query
- Refactor the layout to hide the storybook main area when the route doesn't match (as apposed to the exception `/settings` was before)

## How to test

- create a sandbox
- add a custom manager entry
- here's the code I've been using to test with:
  ```tsx
  import type { ComponentProps, FC } from 'react';
  import React, { memo, Fragment } from 'react';
  import { addons, types } from '@storybook/manager-api';
  
  import { Icons, IconButton, Bar, TabButton } from '@storybook/components';
  import { styled } from '@storybook/theming';
  import { Route } from '@storybook/router';
  
  const Toolbar = styled(
    ({ shown = true, ...props }: ComponentProps<typeof Bar> & { shown: boolean }) => (
      <Bar {...props} />
    )
  )(
    {
      position: 'absolute',
      left: 0,
      right: 0,
      top: 0,
      transition: 'transform .2s linear',
    },
    ({ shown }) => ({
      transform: shown ? 'translateY(0px)' : 'translateY(-40px)',
    })
  );
  
  const FrameWrap = styled.div<{ offset: number }>(({ offset }) => ({
    position: 'absolute',
    overflow: 'auto',
    left: 0,
    right: 0,
    bottom: 0,
    top: offset,
    zIndex: 3,
    transition: 'all 0.1s linear',
    height: `calc(100% - ${offset}px)`,
    background: 'transparent',
  }));
  
  const Centered = styled.div({
    position: 'relative',
  
    width: '100%',
    height: '100%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  });
  
  addons.register('@storybook/addon-debugger', (api) => {
    addons.add('@storybook/addon-debugger/tool', {
      title: 'debugger!',
      type: types.TOOLEXTRA,
      match: ({ viewMode }) => !!(viewMode && viewMode.match(/^(story|docs)$/)),
      render: () => (
        <Fragment>
          <IconButton
            key="debugger"
            title="navigate to debugger-view"
            onClick={() =>
              api.navigateUrl(`/debugger/${api.getCurrentStoryData().id}`, { plain: false })
            }
          >
            <Icons icon="lightning" />
          </IconButton>
        </Fragment>
      ),
    });
  
    const DebuggerPage: FC = () => {
      return (
        <Route path="/debugger" startsWith>
          <Toolbar shown border>
            <TabButton active>A tab</TabButton>
            <IconButton
              key="first"
              title="Go to first story, for some reason. It's just a demo."
              onClick={() => api.selectFirstStory()}
            >
              <Icons icon="star" />
            </IconButton>
          </Toolbar>
          <FrameWrap offset={40}>
            <Centered>
              This is the contents of my addon, in a full viewport experience, what a joy!
            </Centered>
          </FrameWrap>
        </Route>
      );
    };
  
    addons.add('@storybook/addon-debugger/panel', {
      title: 'Debugger',
      type: types.experimental_PAGE,
      url: '/debugger',
      render: memo(DebuggerPage),
    });
  });
  ```
- navigate to `/`: expect the story to be shown
- navigate to `/settings/about` : expect the about page to be shown
- navigate to `/debugger` : expect the debugger panel to be shown
- click on the ⭐ icon in the toolbar (whilst on the `debugger` page : expect the route to change, and the first story to be shown
- use the browser back button : expect the debugger page to be shown again
- use the browser forward button : expect the story to be shown again
- click the ⚡ icon on the right side of the toolbar (whilst on the canvas) : expect the route to change and the debugger page to be shown

I have done the best I can to have some of this also make sense for the mobile view, but this is really up for grabs. We need to determine what to do regarding pages for mobile.

## Checklist

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to make sure your PR is ready to be reviewed. -->

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->

ndelangen committed:
add a new addon type: Page, refactor manager layout to allow for pages

ndelangen committed:
apply router changes from https://github.com/storybookjs/storybook/pull/23292

ndelangen committed:
make `/` the default path, even when there's no query in the URL

ndelangen committed:
cleanup

ndelangen committed:
fix an ancient bug where navigating before preview load, the preview ignores your selection

ndelangen committed:
Merge branch 'norbert/ui-tabs-types-improvements' into norbert/page-addons-refactor

ndelangen committed:
cleanup

ndelangen committed:
fix

ndelangen committed:
fix

ndelangen committed:
Merge branch 'norbert/ui-tabs-types-improvements' into norbert/page-addons-refactor

ndelangen committed:
Merge branch 'norbert/ui-tabs-types-improvements' into norbert/page-addons-refactor

ndelangen committed:
Merge branch 'norbert/page-addons-refactor' of github.com:storybookjs/storybook into norbert/page-addons-refactor

ndelangen committed:
Merge branch 'norbert/ui-tabs-types-improvements' into norbert/page-addons-refactor

ndelangen committed:
Merge branch 'norbert/ui-tabs-types-improvements' into norbert/page-addons-refactor

ndelangen committed:
Merge branch 'norbert/ui-tabs-types-improvements' into norbert/page-addons-refactor

ndelangen committed:
Merge branch 'release/7.2' into norbert/page-addons-refactor

ndelangen committed:
Merge branch 'release/7.2' into norbert/page-addons-refactor

ndelangen committed:
rename

ndelangen committed:
cleanup

ndelangen committed:
I like tests

ndelangen committed:
Merge branch 'release/7.2' into norbert/page-addons-refactor

tmeasday commented on code/lib/manager-api/src/index.tsx:
```diff
@@ -419,36 +419,41 @@ const addonStateCache: {
     };
     const stateInitializationHandlers = {
-      [SET_STORIES]: () => {
+      [SET_STORIES]: async () => {
```
What's going on with this code @ndelangen? Given we don't fire `SET_STORIES` for SSv7, it seems weird to have code that relies on it.

ndelangen replied:
The code handling this event was already there, I just made a change to ensure `await` setting state before emitting the event.

`useSharedState` can be used with SSv6 as well.

tmeasday replied:
I know I guess I just meant while you were looking at it maybe we should consider removing it given it doesn't run in ssv7 and I guess things still work?

ndelangen replied:
@tmeasday I'll leave the code alone for now, and we'll remove SSv6 hopefully in 8.0 🤞

tmeasday commented on code/lib/manager-api/src/modules/addons.ts:
```diff
@@ -18,11 +19,19 @@ export interface SubAPI {
   /**
    * Returns a collection of elements of a specific type.
    * @protected This is used internally in storybook's manager.
+   * @template FAKE - The type of the elements in the collection.
```
What's going on here?

ndelangen replied:
`getElements` was already generic, but I want to ignore that when passed in, because after my changes it's usually wrong.

Now it will return the correct type if you pass the correct `type` argument into it.
So to make `getElements` backwards compatible I discard the first type-argument.
I felt like being on the safe side of backwards compatibility here.

I could remove it and see if it's really needed.
`getElements` is sort-of an internal API only.

tmeasday replied:
I guess making a breaking change isn't kosher but maybe ok to make a breaking type change in this case? Make make it `| any` and drop it if it doesn't match the new type?

tmeasday commented on code/lib/manager-api/src/modules/stories.ts:
```diff
@@ -600,13 +601,25 @@ export const init: ModuleFn<SubAPI, SubState, true> = ({
+           * - If the user has navigated away already.
+           * - If the user started storybook with a specific page-URL like "/settings/about"
+           */
+          if (state.path === '/' || state.viewMode === 'story' || state.viewMode === 'docs') {
+            if (
+              state.viewMode &&
+              state.storyId &&
+              state.viewMode !== viewMode &&
+              state.storyId !== storyId
+            ) {
+              fullAPI.emit(SET_CURRENT_STORY, { storyId: state.storyId, viewMode: state.viewMode });
+            } else if (state.storyId !== storyId || state.viewMode !== viewMode) {
+              navigate(`/${viewMode}/${storyId}`);
+            }
```
I think this code could be a lot clearer if some of these checks were pulled out into variables, like:

```
const stateHasSelection = state.viewMode && state.storyId;
const stateSelectionDifferent = state.storyId !== storyId || state.viewMode !== viewMode;
```

tmeasday commented on code/lib/manager-api/src/modules/stories.ts:
```diff
@@ -600,13 +601,25 @@ export const init: ModuleFn<SubAPI, SubState, true> = ({
+            if (
+              state.viewMode &&
+              state.storyId &&
+              state.viewMode !== viewMode &&
+              state.storyId !== storyId
```
I think maybe this should be an `||`? But this code is very confusing right now. I appreciate the comment but it's still not explaining what the nested ifs are doing.

ndelangen replied:
You're right, This took a few iteration to get right, so the comment explaining it should be even clearer.

tmeasday commented on code/lib/preview-api/src/modules/preview-web/PreviewWithSelection.tsx:
```diff
@@ -220,9 +220,15 @@ export class PreviewWithSelection<TFramework extends Renderer> extends Preview<T
   }
 
   async onSetCurrentStory(selection: { storyId: StoryId; viewMode?: ViewMode }) {
+    /**
+     * set state immediately.
+     * If the store is still in the init phase, we'll read it from the URL.
+     * But we might be receiving a SET_CURRENT_STORY event and we should load that story when we init.
+     */
```
```suggestion
    /**
     * At the end of the initialization promise we will read the current story from the selection store,
     * so make sure we've updated it with the new selection or we'll lose track of it at the end of init.
     */
```

Nice pickup! Do I have this right?

We should add a test for this scenario in `PreviewWeb.test.ts`. We can pair on it if you want.

tmeasday commented on code/lib/types/src/modules/addons.ts:
```diff
@@ -339,6 +339,39 @@ interface FCWithoutChildren<P = {}> {
+   * The unique id of the page.
+   */
+  id: string;
+  /**
+   * The URL to navigate to when Storybook needs to navigate to this page.
+   */
+  url: string;
```
Don't we want to support a URL scheme? Like I am imagining we'll want to have something like `/visual-test/$storyId` (or similar containing `$storyId`). Perhaps what I am saying is that it should maybe be possible to add a view mode?

tmeasday commented:
It worked well but I did see these warnings:

![image](https://github.com/storybookjs/storybook/assets/132554/8c78ff38-bedd-4f09-aebc-7d113c92d03e)

ndelangen committed:
Update code/lib/preview-api/src/modules/preview-web/PreviewWithSelection.tsx

Co-authored-by: Tom Coleman <tom@chromatic.com>

ndelangen committed:
remove the fake generic type argument

ndelangen committed:
improve readability of api's stories module init

ndelangen committed:
fix cyclical state setting

ndelangen committed:
addon render function are now called like React elements, thus keys are no longer needed.

This is not a breaking change, just logs a warning from react as far as i can see

ndelangen committed:
fix type issues

shilman reviewed (commented):
During the review you mentioned that there is a breaking TS change where `Addon_Type`'s `type` field is now required. Please add this to `MIGRATION.md`.

I love the use of `experimental_` in the name. Playwright also does this and I think it's a great convention -- we should do more.

ndelangen committed:
Merge branch 'release/7.2' into norbert/page-addons-refactor

tmeasday commented on code/lib/manager-api/src/modules/stories.ts:
```diff
@@ -600,13 +601,26 @@ export const init: ModuleFn<SubAPI, SubState, true> = ({
+           * - If the user has navigated away already.
+           * - If the user started storybook with a specific page-URL like "/settings/about"
+           */
+          if (isCanvasRoute) {
+            if (stateHasSelection && stateSelectionDifferent) {
+              // The manager state is correct, the preview state is lagging behind
+              fullAPI.emit(SET_CURRENT_STORY, { storyId: state.storyId, viewMode: state.viewMode });
+            } else if (stateSelectionDifferent) {
+              // The preview state is correct, the manager state is lagging behind
+              navigate(`/${viewMode}/${storyId}`);
+            }
```
So much easier to understand now 👏

tmeasday commented on code/lib/preview-api/src/modules/preview-web/PreviewWithSelection.tsx:
```diff
@@ -220,9 +220,14 @@ export class PreviewWithSelection<TFramework extends Renderer> extends Preview<T
   }
 
   async onSetCurrentStory(selection: { storyId: StoryId; viewMode?: ViewMode }) {
+    /**
+     * At the end of the initialization promise we will read the current story from the selection store,
+     * so make sure we've updated it with the new selection or we'll lose track of it at the end of init.
+     */
+    this.selectionStore.setSelection({ viewMode: 'story', ...selection });
+
     await this.storyStore.initializationPromise;
```
I'd still like a unit test for this change

ndelangen replied:
added!

ndelangen committed:
Merge branch 'release/7.2' into norbert/page-addons-refactor

ndelangen committed:
Merge branch 'norbert/page-addons-refactor' of github.com:storybookjs/storybook into norbert/page-addons-refactor

ndelangen committed:
add test for early setting of selection in preview-web

ndelangen committed:
add migration documentation

ndelangen commented:
@shilman I don't think the `Addon_BaseType` needs to be documented in the migration guide.. it is only relevant if you're consuming registered addon, not when injecting them.

I did end up writing some migration guide bits for the increased strictness of the manager-api

tmeasday commented on code/lib/preview-api/src/modules/preview-web/PreviewWeb.test.ts:
```diff
@@ -1552,6 +1552,32 @@ describe('PreviewWeb', () => {
       jest.useRealTimers();
     });
 
+    it('listens for events during setup', async () => {
```
Maybe put this next to this one? it's a minor variant on it -- did you check it was failing before?

https://github.com/storybookjs/storybook/blob/428e6ce087c680c91c92a2eac74034069cb9dd72/code/lib/preview-api/src/modules/preview-web/PreviewWeb.test.ts#L1629

ndelangen replied:
I'm confused. No that story wasn't failing, but it seems to test for the very same thing.

tmeasday replied:
Slightly different in that the existing one didn't have a selection before init.

ndelangen committed:
move test

ndelangen committed:
Merge branch 'release/7.2' into norbert/page-addons-refactor

tmeasday committed:
Straighten out behaviour when messages are received during init

ndelangen committed:
fix

ndelangen committed:
Merge branch 'release/7.2' into norbert/page-addons-refactor
